File: all-submissions/0207-course-schedule/solution.java
class Solution {
    
    Set<Integer> seen = new HashSet<>();
    Set<Integer> stack = new HashSet<>(); 

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, Set<Integer>> preReqMap = new HashMap<>();
        for(int i=0;i<prerequisites.length;i++){
            int course = prerequisites[i][0];
            int prereq = prerequisites[i][1];
            if(!preReqMap.containsKey(course)){
                preReqMap.put(course, new HashSet<>());
                
            }
            preReqMap.get(course).add(prereq);
        }
        for(int course: preReqMap.keySet()) {
            if(!seen.contains(course)){
                if (isPrereqCyclic(course, preReqMap)){
                    return false;
                }
            }
        }
        return true;
    }
    
    public boolean isPrereqCyclic(int course, Map<Integer, Set<Integer>> preReqMap) {
        seen.add(course);
        stack.add(course);

        for (Integer neighbor : preReqMap.getOrDefault(course, Collections.emptySet())) {
            if (!seen.contains(neighbor)) {
                if (isPrereqCyclic(neighbor, preReqMap)) {
                    return true;
                }
            } else if (stack.contains(neighbor)) {
                // Found a back edge → cycle
                return true;
            }
        }

        stack.remove(course);  // backtrack
        return false;
    }
}

File: all-submissions/0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        adj = defaultdict(list)

        for prereq in prerequisites:
            adj[prereq[1]].append(prereq[0])
            indegree[prereq[0]] +=1

        queue = deque()
        visited = 0
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        while(queue):
            course = queue.popleft()
            visited +=1
            for c in adj[course]:
                indegree[c] -=1
                if indegree[c] == 0:
                    queue.append(c)

        return visited == numCourses


